_posts/2019-01-04-nextjs.md

```diff
@@ -0,0 +1,290 @@
+---
+layout: post
+title: Create Universal Apps Using Next.js
+date: 2019-03-18 10:00:00 +0200
+categories: javascript
+author_names: ["Andreas Geier", "Sergii Paryzhskyi"]
+read_time : 10
+show_related_posts: false
+square_related:
+feature_image: posts/2019-01-04-nextjs/poster.png
+---
+
+Let's talk at first about libraries for creating universal applications. On Github you can find a large number of projects implementing the idea of universal web applications. However, all these projects have common drawbacks:
+
+- missing functionality such as server side rendering (SSR) or i.e. hot reloading
+- these are projects for quick start, not a library
+- small number of project contributors
+- projects were not updated when new versions of react.js were released
+
+The first successful solution was the Next.js library, which, has 3600 contributors and over 33K “stars” on Github. To evaluate the benefits of this library, consider what kind of functionality you need to provide for the operation of a universal web application:
+
+- Asynchronous data loading. In order for the results of asynchronous requests to be received prior to server rendering, Next.js implements a special type of the page component, which has an asynchronous static async getInitialProps ({req}) life cycle event.
+
+- Server-side rendering (SSR).
+
+- Passing server component state to client.
+
+- Creating a component on the side of a web browser and linking it to an HTML document.
+
+- Routing on the server and on the client should also be universal. That is, the same definition of routing should work for both server and client code.
+
+- Code splitting. For each page, only the necessary JavaScript code should be loaded, not the entire application. When moving to the next page, the missing code should be reloaded - without reloading the same modules, and without unnecessary modules.
+
+All these tasks are successfully solved by the Next.js library. This library is based on a very simple idea. It is proposed to introduce a new type of component - “page”, in which there is an async static method async getInitialProps ({req}). The “page” component is a normal React component. This type of component can be thought of as a new type in the series: “component”, “container”, “page”.
+
+## Sample Project
+
+Let's create a simple project with Next.js to demonstrate how it works. You will be surprised how little we need to do in order to start our actual application. Some complexity is hidden by the library and the rule convention over configuration is helping to keep code on our side really slim.
+
+Create a new directory for our sample project `mkdir sample-app && cd sample-app`. Generate an empty npm project here with `npm init -y`. Then install Next.js together with React dependencies: `npm i next react react-dom --save`.
+
+Now we just have to add a new npm run script to our `package.json`:
```

Comment:
we -> you

I mean ... assuming you are writing to your reader, and not yourselves ;)